gslib/utils/shim_util.py

```diff
@@ -0,0 +1,219 @@
+from __future__ import division
+from __future__ import unicode_literals
+
+import collections
+import os
+import subprocess
+
+from boto import config
+from gslib import exception
+
+VALID_USE_GCLOUD_STORAGE_VALUES = (
+    'never',
+    'if_available_else_skip',
+    'always',
+    'dry_run',
+)
+
+
+class GcloudStorageFlag(object):
+
+  def __init__(self, gcloud_flag, supports_output_translation=False):
+    """Initializes GcloudStorageFlag.
+    
+    Args:
+      gcloud_flag (str): The name of the gcloud flag.
+      support_output_translation (bool): If True, this flag in gcloud storage
+        supports printing gsutil formatted output.
+    """
+    self.gcloud_flag = gcloud_flag
+    self.supports_output_translation = supports_output_translation
+
+
+class GcloudStorageMap(object):
+  """Mapping to translate gsutil command to its gcloud storage equivalent."""
+
+  def __init__(self,
+               gcloud_command,
+               flag_map,
+               supports_output_translation=False):
+    """Intalizes GcloudStorageMap.
+    
+    Args:
+      gcloud_command (dict|str): The corresponding name of the command to be
+        called in gcloud. If this command supports sub-commands, then this 
+        field must be a dict of sub-command-name:GcloudStorageMap pairs.
+      flag_map (dict): A dict of str to GcloudStorageFlag. Mapping of gsutil
+        flags to their equivalent gcloud storage flag names.
+      supports_output_translation (bool): Indicates if the corresponding
+        gcloud storage command supports the printing gsutil formatted output.
+    """
+    self.gcloud_command = gcloud_command
+    self.flag_map = flag_map
+    self.supports_output_translation = supports_output_translation
+
+
+def _get_gcloud_binary_path():
+  cloudsdk_root = os.environ.get('CLOUDSDK_ROOT_DIR')
+  if cloudsdk_root is None:
+    raise exception.GcloudStorageTranslationError(
+        'Gcloud binary path cannot be found.')
+  return os.path.join(cloudsdk_root, 'bin', 'gcloud')
+
+
+class GcloudStorageCommandMixin(object):
+  """Provides gcloud storage translation functionality.
+  
+  The command.Command class must inherit this class in order to support
+  converting the gsutil command to it's gcloud storage equivalent.
+  """
+  # Mapping for translating gsutil command to gcloud storage.
+  gcloud_storage_map = None
+
+  def __init__(self):
+    self._translated_gcloud_storage_command = None
+    self._translated_env_variables = None
+
+  def _get_gcloud_storage_args(self, sub_opts, gsutil_args, gcloud_storage_map):
+    if gcloud_storage_map is None:
+      raise exception.GcloudStorageTranslationError(
+          'Command "{}" cannot be translated to gcloud storage because the'
+          ' translation mapping is missing.'.format(self.command_name))
+    args = []
+    if isinstance(gcloud_storage_map.gcloud_command, str):
+      args = gcloud_storage_map.gcloud_command.split()
+    elif isinstance(gcloud_storage_map.gcloud_command, dict):
```

Comment:
Why do we have multiple types for `gcloud_command` (`str` and `dict`)?

Reply:
It can be a string if there is one-to-one mapping. But for cases which has sub-commands, we use a dict, because a sub-command will have its own mapping. See the test case test_get_gcloud_storage_args_parses_subcommands or the design doc for more info.